### PR TITLE
Fix some issues with the Test Game and add some missing tests

### PR DESCRIPTION
**Pull request by Mimigris**
This pull request was made to fix some issues present in the 2k Test Game, in addition to adding some missing tests to better test the differences between EasyRPG Player and the RPG_RT.
Due to the addition and edition of some text lines, this pull request will impact a bit on Pull Request # 37 of the Test Game, but it should not cause an issue ultimately since the lines can easily be updated thanks to lcftrans.

**_Content changed in the RPG 2000 Test Game_**

- Updated the IRC Bot text in the _Standard World Test_ to list it as EV0002 instead of 0001 and listed libera.chat instead of freenode.net since it is no longer available
- Added a _Stress Test_ as requested on the [wiki page of the EasyRPG Wiki](https://wiki.easyrpg.org/development/rpg-maker-test-suite?rev=1326916631) (not perfect so feel free to request changes/submit suggestions)
- Added a todo list on the _Map Events_ map (current list is the two issues currently present in the Test Game, + make a better Stress Test than mine)
- Added a Set Vehicle Position: Player in _Vehicle Commands_ (RPG_RT exploit used by some games)
- Added a Wait while a message is displayed test in _Wait_ (linked to Player issue #2890)
- Added two events to properly test fadein and fadeout of a track in _Music Playing_ (linked to Player issue #2892)
- Added a test to play an ogg track in _Music Playing_ (RPG_RT can't play it, but EasyRPG can)
- Added a testcase in _Pan Screen_ to test the layering issues present in a looping map in RPG_RT
- Added an event to remove each member of your party in _Status_ to be able to properly test the Death Status
- Added two known edge-cases to _Image Operations_: call a file from another folder as a picture (../System/System.png here) and call an empty filename as a picture (I didn't got an 100% accurate result here)
- Replaced the default Battle Start sound to be Bell instead of Holiness1 since it does not exist
- Fixed the script with the event Jumping to label 5 times who had the mention FIXME in _Loop and Goto/Label_
- Fixed an issue where the two events to test the loop played the same music in _Music Playing_
- Fixed an issue in _Shake Screen_ and _Memorize/Go to Place_ where using the stairs in the second map would need to press the action button instead of just walking on it, and where the game would put you out of bounds due to the use of an outdated script
- Fixed an issue in _Memorize/Go to Place_ where getting your coordinates would display a static message instead of your actual coordinates, making them not accurate depending on the side chosen
- Added a reset to default vehicles chara in _Change Vehicles Chara_
- Fixed an issue where moving the vehicles to the main map in _Vehicles commands_ would put some vehicles only accessible from the bridge that teleport you, making them inaccessible
- Made the message settings be forced to be displayed below when in _Tile substitution_ to be able to always see the tiles changed (the message settings is reset to default on map exit)
- Fixed a lock that could occur when talking to some events in _Event Movement_ from specific sides, or when talking to some events at a specific moment
- Redone the Frequency test event in _Event Movement_ since it was not very visible, and added skip if cannot move to said NPC to make it be properly viewed
- Fixed an issue in _Branching_ where changing Alex's HP to test the Branch would cause a Game Over since his HP were decreased when doing so, and that a Game Over was listed as possible
- Fixed an issue in _Branching_ where Alex's level couldn't be set as 50
- Added a return to origin and unfix on map exit of _Pan Screen_ to properly be able to move on to the next test
- Fixed an issue where a line was duplicated by error in _Music Playing_
- Added more heroes on entrance of _Change System Sounds_ to make it possible to listen the "Touched!" sound in fight
- Made the battle escapable in _Change System Sounds_ to be able to hear the escape sound
- Fixed an issue where Success1.wav was incorrectly listed as Success.wav in _Change System Sounds_, and where the Buzzer sound was incorrectly reset as Buzzer1 instead of Item2
- Fixed an issue in _Change System BGM_ where the Vehicle 2 music would be changed to the exact same track, making it impossible to test
- Fixed an issue in _Change System BGM_ where changing the BGM Battle would change to the default Battle music used during fights, making it weird to be tested
- Fixed an issue in _Change System BGM_ where the Battle End, Game Over and Inn tracks were incorrectly reset when choosing said option
- Fixed an issue in the _Items Shop_ where the Test Item was incorrectly mentioned as Tear Drop
- Made Brian and Alex status reset on map exit of _Change hero chara/face_
- Removed an incorrect line in _Hero Operation_ saying that removing a null member from your party would crash RPG_RT while it does not seems to be case
- Added a text in _Equip_ to mention that Bow can't be equipped to Carol and Deiji when the game attempts to equip it to everyone
- Moved the line of X to the correct placement in _Show Face_ to make the test be more accurate
- Fixed an issue in _Choices_ where a disallow choice was missing for one of the tests
- Fixed an issue in _Timer Operations_ where the timing command and the message used were incorrect
- Fixed a line that couldn't be displayed entirely in _Money Operation_
- Fixed a lot of typos (purpose misspelled as prupose in _Vehicles commands_, myself and hero as myself an hero in _Hero Flash_, frequency as frecuency in _Event Movement_, sec (from seconds) as seg in _Screen Flash_, borders will be shown as bordes will shown and speed as vel in _Pan Screen_, poison as posion in _Status_, because as beacuse in _Timer Operations_, Choice 1 has been choosen as Choice 1 has choosen in _Choices_ and some that I forgot)

Listed the following tests as working entirely:
- _Message Options_ (continuing events)
- _Timer Operation_
- _Experience_ (RPG_RT quirks)
- _Level_ (RPG_RT quirks)
- _Parameters_ (Adding/Removing negative ATK)
- _HP/MP recover/remove_ (Adding/Removing negative HP)
- _Status_ (Death Status)
- _Damage_
- _Change vehicles chara_
- _Change system BGM_
- _Change system sounds_
- _Change system (graphic)_
- _Change default screen transitions_
- _Inn_
- _Item Shop_
- _Branching_
- _Vehicle Commands_
- _Screen Tone_
- _Shake Screen_
- _Pan Screen_
- _Weathers Effect_
- _Battle animations_
- _Event Movement_
- _Menu Testing_
- _Teleport and Escape_
- _Swap events_
- _Get terrain ID_
- _Get event ID_
- _Key input_
- _Loop and Goto/Label_
- _Event deletion and cancellation_
- _Tile substitution_
- _Call Event_

Listed the following tests as working except the following issues:
- _Transitions_ EXCEPT the mentioned issues at the end of the pull request
- _Equip_ EXCEPT Equip to members who can't
- _Image Operations_ EXCEPT empty filename
- _Wait_ EXCEPT Wait during message
- _Music playing_ EXCEPT balance and fadein/out
- _Panorama_ EXCEPT display on looping map

**_Content changed in the RPG 2003 Test Game_**

- Moved the Test Game from the unofficial RTP to the official RTP to make it easier to test (I've not done that for the RPG 2000 Test Game since it is more complex)
- Moved the tests about added features in 2k3 from the 2k Test Game to the 2k3 Test Game for obvious reasons
- Expanded the TestGame by adding several events only available in 2k3, a test for the _Generate Dungeon_ feature and some tests linked to battles (like calling a Common Event in a battle)

**_Issues found while testing this:_**

- Setting Parallax on a looping map does not seems to work on a looping map in the continuous build (reported in Player #2889)
- Timing issue for fadein and fadeout when playing mp3, wav or ogg music (reported in Player #2892)
- Wrong order for the teleport points (RPG_RT seems to sort by map ID while EasyRPG Player does it by order of addition; since this is not even saved in the savefile EasyRPG Player will ultimately use the same order as the default RPG_RT on the next load of the save)
- Call an empty filename picture does not seems to be entirely accurate - needs more testing
- The transition Show: Inwards is not supported
- The transition Show: Fade in is too fast compared to the RPG_RT
- Removing an invalid hero through Hero Operation does not seems to cause an issue on RPG_RT, while Easy RPG Player displays a warning message
- Add Equipment can be used to equip an item to an hero that can't equip it (e.g. you can equip a bow to a character that should not be able to equip it)
- The Encounter Rate seems to not be accurate (e.g. with an Encounter Rate of 1 step, you can not trigger a battle while walking, while this is not the case in RPG_RT)
- Teleport/Escape seems to not remove the pictures currently displayed on-screen on the RPG_RT; needs more research

The following commands called in battle in 2k3 through a Common Event don't work correctly:
- Enemy Encounter
- Shop Process
- Inn Message
- Enter Hero Name
- Transfer Player
- Move to Variables
- Get On/Off Vehicle
- Set Vehicle Position (to Player only, the other vehicles work)
- Hide Screen
- Show Screen
- Pan Screen
- Show Battle Animation
- Flash Hero
- Wait for All Movement
- Stop All Movement
- Open Save Menu
- Open Menu
- Open Load Menu
- Toggle ATB Mode

---
- Wait: key input and Key input: wait in fight will lock the game on RPG_RT, but Easy RPG Player support them like outside of a battle
- Wait during a fight will stop the heroes from moving in RPG_RT (they will stay still) but they continue to move in EasyRPG Player
- Force Escape: All Enemies on the start of a fight will partially display the fight menu
in RPG_RT (only the monsters list will start to be displayed and their names will not be displayed) while EasyRPG Player will display the menus entirely
- You can get interrupted while moving in the Fight Menu (e.g. a monster is attacking while you were moving), causing a display issue
graphic_issue_screenshot
- Change Enemy HP: Decrease more HP than the total HP that they still have without allowing to kill them will display the HP that were asked to be removed in EasyRPG Player, while RPG_RT display the actual HP removed
- The Level Up message is not correctly displayed while in fight (RPG_RT displays the message in the 1 line message box, while EasyRPG Player displays it in the 4 lines message box). Additionally, the level up message of the v1.11 of RPG Maker 2003 should normally be "(Character) <custom line> (Level)(X)!" on RPG_RT, but EasyRPG Player seems to forget the ! at the end "(Character) <custom line> (Level)(X)".
image
image

Issues that are extremely minor/only useful if you want to be 1:1 with the RPG_RT:
- Screen Tone directly go from one color to another while EasyRPG is more consistent
- Pan Screen on a looping map causes some display issues linked to the layering of the events (some events will be above/under the player and some events will disappear), not replicated on EasyRPG Player
- Money Operations, Experience and Level overflow and underflow issues are not replicated
- Adding an invalid item/skill to your inventory and opening it causes a crash on RPG_RT, while ERPG entirely ignores this without a warning message
- Wait for key input seems to support Decision key + keys labeled from 10 to 24 on RPG_RT, but only supports the Decision key in EasyRPG Player
- The message box in fight can be set as invisible in EasyRPG Player, while this is not the case in RPG_RT where the message box is forced to be visible
- A timer stopping a battle removes the message currently displayed in EasyRPG Player, while RPG_RT keeps the message (same thing for Choices, but the choice chosen will have no effect if selected outside of the battle)
---
Feel free to request changes for this pull request or ask questions.

### CODE REVIEW

**Review by fdelapena — APPROVED** [review]

**Comment by Ghabry** [comment]
Oh wow. Thanks for this. I will slowly fight through the list 😅

**Comment by Ghabry** [comment]
@Mimigris 

Issues I found:

- In Show Message the top right event appears to work in EasyRPG Player but is not marked with a white tile
- The bot is still called EV0001 in IRC chat. Only the version number is EV0002.
- Brian and Deiji use now RTP Facesets instead of our "large Charsets"

I'm honestly a bit overwhelmed by your work and the issues you found. Except for the panorama one they do not look like regressions or game breaking so is likely fine to omit them from 0.8.0.

---------

Sorry for generating even more work but can you create issues in Player for them?

When ~~striked through~~ no issue needed

- ~~Wrong order for the teleport points~~ see patch below
- ~~Call an empty filename picture does not seems to be entirely accurate - needs more testing~~ As this is abusing an implementation detail being 100% accurate is not important imo (unless it breaks games)
- **Combine into one issue**:
  - The transition Show: Inwards is not supported
  - The transition Show: Fade in is too fast compared to the RPG_RT
- ~~Removing an invalid hero through Hero Operation does not seems to cause an issue on RPG_RT, while Easy RPG Player displays a warning message~~ I will quickly patch this
- **Create issue**: Add Equipment can be used to equip an item to an hero that can't equip it
- **Create issue**: The Encounter Rate seems to not be accurate (e.g. with an Encounter Rate of 1 step, you can not trigger a battle while walking, while this is not the case in RPG_RT)
- **Create issue**: Teleport/Escape seems to not remove the pictures currently displayed on-screen on the RPG_RT
- **Create one issue**: All the battle stuff ;)

Teleport target fix:

```diff
diff --git a/src/game_targets.cpp b/src/game_targets.cpp
index da5b5cefd..13307625d 100644
--- a/src/game_targets.cpp
+++ b/src/game_targets.cpp
@@ -40,6 +40,9 @@ void Game_Targets::AddTeleportTarget(int map_id, int x, int y, bool switch_on, i
 	iter->map_y = y;
 	iter->switch_on = switch_on;
 	iter->switch_id = switch_id;
+
+	// Teleports must be sorted by map id.
+	std::sort(teleports.begin(), teleports.end(), [](auto& l, auto& r) { return l.map_id < r.map_id; });
 }
 
 void Game_Targets::RemoveTeleportTarget(int map_id) {
```
